Mathlib/Algebra/Algebra/Subalgebra/Unitization.lean

```diff
@@ -12,86 +12,162 @@ import Mathlib.Algebra.Star.NonUnitalSubalgebra
   by_cases hr : r = 0
-  · simp only [hr, map_zero, Unitization.inl_zero, zero_add] at hx' ⊢
-    rw [← ZeroMemClass.coe_zero s] at hx'
-    exact congr_arg _ (Subtype.coe_injective hx')
-  · refine' False.elim (h1 _)
-    rw [← eq_sub_iff_add_eq, zero_sub] at hx'
-    replace hx' := congr_arg (fun y => r⁻¹ • y) hx'
-    simp only [Algebra.algebraMap_eq_smul_one, ← smul_assoc, smul_eq_mul, inv_mul_cancel hr,
-      one_smul] at hx'
-    exact hx'.symm ▸ SMulMemClass.smul_mem r⁻¹ (neg_mem a.prop)
+  · ext <;> simp [hr] at hx ⊢
+    exact hx
+  · exact (h r hr <| hx ▸ (neg_mem a.property)).elim
+
+/-- This is a generic version which allows us to prove both
+`NonUnitalSubalgebra.unitization_injective` and `NonUnitalStarSubalgebra.unitization_injective`. -/
+theorem AlgHomClass.unitization_injective {F R S A : Type _} [Field R] [Ring A]
+    [Algebra R A] [SetLike S A] [hSA : NonUnitalSubringClass S A] [hSRA : SMulMemClass S R A]
+    (s : S) (h1 : 1 ∉ s) [AlgHomClass F R (Unitization R s) A] (f : F)
+    (hf : ∀ x : s, f x = x) : Function.Injective f := by
+  refine AlgHomClass.unitization_injective' s (fun r hr hr' ↦ ?_) f hf
+  rw [Algebra.algebraMap_eq_smul_one] at hr'
+  exact h1 <| inv_smul_smul₀ hr (1 : A) ▸ SMulMemClass.smul_mem r⁻¹ hr'
+
+variable {R S A : Type _} [Field R] [Ring A] [Algebra R A]
+  [SetLike S A] [hSA : NonUnitalSubringClass S A] [hSRA : SMulMemClass S R A] (s : S)
+
+theorem NonUnitalSubalgebra.unitization_injective (h1 : (1 : A) ∉ s) :
+    Function.Injective (NonUnitalSubalgebra.unitization s) :=
+  AlgHomClass.unitization_injective s h1 (NonUnitalSubalgebra.unitization s) fun _ ↦ by simp
 
 /-- If a `NonUnitalSubalgebra` over a field does not contain `1`, then its unitization is
 isomorphic to its `Algebra.adjoin`. -/
-noncomputable def NonUnitalSubalgebra.unitizationAlgEquiv {R S A : Type _} [Field R] [Ring A]
-    [Algebra R A] [SetLike S A] [NonUnitalSubringClass S A] [SMulMemClass S R A]
-    (s : S) (h1 : (1 : A) ∉ s) : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A) :=
-  AlgEquiv.ofBijective (NonUnitalSubalgebra.unitization s)
-    ⟨NonUnitalSubalgebra.unitization_injective s h1, NonUnitalSubalgebra.unitization_surjective s⟩
+@[simps!]
+noncomputable def NonUnitalSubalgebra.unitizationAlgEquiv (h1 : (1 : A) ∉ s) :
+    Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A) := by
+  let algHom : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A) :=
+    ((NonUnitalSubalgebra.unitization s).codRestrict _
+      fun x ↦ NonUnitalSubalgebra.unitization_range s ▸ ⟨x, rfl⟩)
+  refine AlgEquiv.ofBijective algHom ⟨?_, fun x ↦ ?_⟩
+  · have := AlgHomClass.unitization_injective s h1
+      ((Subalgebra.val _).comp <| algHom) fun _ ↦ by simp
+    simp only [AlgHom.coe_comp, Subalgebra.coe_val] at this
+    exact this.of_comp
+  · obtain (⟨a, ha⟩ : (x : A) ∈ (NonUnitalSubalgebra.unitization s).range) :=
+      (NonUnitalSubalgebra.unitization_range s).symm ▸ x.property
+    exact ⟨a, Subtype.ext ha⟩
```

Comment:
```suggestion
    Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A) :=
  let algHom : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A) :=
    ((NonUnitalSubalgebra.unitization s).codRestrict _
      fun x ↦ (NonUnitalSubalgebra.unitization_range s).le <| AlgHom.mem_range_self _ x)
  AlgEquiv.ofBijective algHom <| by
    refine ⟨?_, fun x ↦ ?_⟩
    · have := AlgHomClass.unitization_injective s h1
        ((Subalgebra.val _).comp <| algHom) fun _ ↦ by simp
      simp only [AlgHom.coe_comp, Subalgebra.coe_val] at this
      exact this.of_comp
    · obtain (⟨a, ha⟩ : (x : A) ∈ (NonUnitalSubalgebra.unitization s).range) :=
        (NonUnitalSubalgebra.unitization_range s).ge x.property
      exact ⟨a, Subtype.ext ha⟩
```

Two changes here:
- no more data defined in tactic mode. It was probably fine but I'm scared of defining data in tactic mode, and I also find it a bit cleaner to have a clear distinction
- I replaced some rewriting along `NonUnitalSubalgebra.unitization_range s` by applying the `.le` or `.ge` version. Personally I think it's a bit cleaner, but if you don't like it I have no problem with the original version.

Reply:
Yeah, I thought the tactic mode was safe because I just used `refine` for the data bits, but this is nicer.